File: cp/internal/api/capabilities.go
package api

// What this control plane can actually be asked for (SIGMA-268).
//
// The dashboard builds its wizard out of the GENERATED catalog
// (web/src/lib/server-catalog.generated.ts), which is the full set of engines
// this codebase knows how to provision. The control plane it is talking to has
// its own, possibly smaller, view: CP_DB_ENGINES / CP_S3_ENGINES cut the list
// (a Postgres-only build is a supported, pre-agreed shape), and store.CreateResource
// refuses anything outside it with `database engine "x" is not enabled on this
// control plane`.
//
// Nothing published that second list, so the wizard could not know it. Both
// halves of the disagreement then failed in the same place and in the worst
// possible way: the operator picked an engine the dialog offered, filled in a
// name, chose a project, an environment and a server, pressed create, and got a
// 422 — after the dialog had closed. The engines are a property of the control
// plane, so the control plane is what has to say them out loud.
//
// Org-scoped and Developer-readable because the wizard is: this is not a
// secret (it is a shape of the deployment, visible in every refusal already),
// but it belongs behind the same token as everything else the dashboard reads,
// and routing it under /v1/orgs/{orgId} keeps it on the one authenticated
// surface rather than inventing a second unauthenticated one.

import (
	"net/http"

	"github.com/Strahinja-Polovina/sigmahub/cp/internal/store"
)

// capabilitiesResponse is deliberately a list of NAMES rather than a rendered
// catalog. The labels, hints and icons are the dashboard's, generated from this
// same Go catalog, so sending them again would be a second copy of the
// vocabulary SIGMA-216 spent its time deleting. What the dashboard cannot
// derive is which of them this deployment turned off.
type capabilitiesResponse struct {
	DBEngines []string `json:"dbEngines"`
	S3Engines []string `json:"s3Engines"`
}

func (s *Server) handleCapabilities(w http.ResponseWriter, r *http.Request) {
	writeJSON(w, http.StatusOK, capabilitiesResponse{
		DBEngines: enabledOrAll(s.dbEngines, store.DBEngineKinds()),
		S3Engines: enabledOrAll(s.s3Engines, store.S3EngineNames()),
	})
}

// enabledOrAll answers with the configured allowlist, or with the whole catalog
// when none was configured.
//
// The fallback matters more than it looks: config.FromEnv already resolves an
// unset CP_DB_ENGINES to the full catalog, so a server built from a real Config
// never takes this branch. A Server built WITHOUT those options — every handler
// unit test, and any future embedding — would otherwise publish "no engines are
// enabled here", and a wizard that believes it would offer nothing at all. An
// absent allowlist means "not restricted", exactly as it does in the store.
func enabledOrAll(enabled, all []string) []string {
	if len(enabled) == 0 {
		return all
	}
	// Copied so a caller cannot mutate the server's configuration through the
	// slice it was handed, and ordered by the catalog rather than by the
	// operator's typing, so two control planes with the same engines enabled
	// answer identically.
	out := make([]string, 0, len(enabled))
	for _, name := range all {
		for _, e := range enabled {
			if e == name {
				out = append(out, name)
				break
			}
		}
	}
	return out
}

File: web/src/server/actions/capabilities.ts
"use server";

// What the control plane behind this dashboard will actually accept (SIGMA-268).
//
// The wizard is built from the generated catalog, which is every engine this
// codebase knows how to provision. A given control plane can be running with a
// smaller set — CP_DB_ENGINES=postgres is a supported, pre-agreed build — and
// its create call refuses the rest with `database engine "x" is not enabled on
// this control plane`. Nothing published that smaller set, so the wizard offered
// the whole catalog and the operator met the disagreement at submit time, with
// the dialog already closed and every decision in it lost.
//
// Demo mode has no control plane to ask and no allowlist to honour: the demo
// implements every engine the catalog defines, so the catalog IS the answer
// there. The same is true when the control plane cannot be reached — see below,
// where a failure deliberately does not narrow the wizard.

import { requireMembership } from "../active-org";
import { DB_ENGINE_CATALOG, S3_ENGINE_NAMES } from "@/lib/server-catalog.generated";
import { cpCapabilities, cpEnabled, type CpCapabilities } from "../cp";

/** Every engine the generated catalog defines — the honest answer whenever no
 *  control plane has narrowed it. */
function catalogCapabilities(): CpCapabilities {
  return {
    dbEngines: Object.keys(DB_ENGINE_CATALOG),
    s3Engines: [...S3_ENGINE_NAMES],
  };
}

/**
 * The engine sets the wizard filters against.
 *
 * Member-visible, like the deploy targets it is shown beside: a developer who
 * cannot read this cannot fill in the form.
 *
 * A control plane that cannot be reached falls back to the full catalog rather
 * than to an empty set, and the direction matters. Narrowing on a failed read
 * would hide engines a working deployment has enabled — turning one bad request
 * into "this product cannot create databases", which is a much worse lie than
 * the one this endpoint exists to fix. Falling back wide degrades to exactly
 * the old behaviour: the create call is still the authority, and it still
 * refuses with a sentence naming the setting.
 */
export async function getEngineCapabilities(orgId: string): Promise<CpCapabilities> {
  await requireMembership(orgId);
  if (!cpEnabled()) return catalogCapabilities();
  try {
    const caps = await cpCapabilities(orgId);
    // A control plane too old to know this route answers 404, which cpFetch
    // throws on; an empty list from a NEW one would mean "nothing is enabled",
    // which no deployment can be — config.FromEnv refuses to boot with an empty
    // allowlist. So treat empty as unknown rather than as a verdict.
    return {
      dbEngines: caps.dbEngines?.length ? caps.dbEngines : catalogCapabilities().dbEngines,
      s3Engines: caps.s3Engines?.length ? caps.s3Engines : catalogCapabilities().s3Engines,
    };
  } catch {
    return catalogCapabilities();
  }
}
